### PR TITLE
Make asset downloading dependent on connection method

### PR DESCRIPTION
**Pull request by pierd**
This PR makes all asset URLs be represented as `ambient-assets:/path/to/asset` and resolved to the actual download URL on client-side. Clients receive base URL for downloading all content in `ServerInfo` transmitted to them during the initial handshake.

### CODE REVIEW

**Comment by chaosprint** [comment]
I guess the `assert::url` should also be modified accordingly, so the `asset_loading` example can work.

**Comment by chaosprint** [comment]
The `asset_loading` example yield following errors:

```
[2023-04-07T10:35:07Z INFO  ambient] Building Asset Loading
[2023-04-07T10:35:10Z INFO  ambient] Done building Asset Loading
[2023-04-07T10:35:10Z INFO  ambient::server] Creating server
[2023-04-07T10:35:10Z INFO  ambient::server] Created server, running at 10.0.0.2:9000
[2023-04-07T10:35:10Z INFO  ambient::server] Starting server
[2023-04-07T10:35:10Z INFO  ambient_network::server] Connecting to proxy server
thread 'tokio-runtime-worker' panicked at 'ambient_std::uncategorized::asset_url::AbsAssetUrl resource doesn't implement the load method', /Users/chaosprint/Dev/AmbientDev/Ambient/crates/asset_cache/src/lib.rs:443:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-04-07T10:35:10Z INFO  ambient_network::client] Connecting to server at 127.0.0.1:9000
[2023-04-07T10:35:10Z INFO  ambient_network::server] [user_1MP7PFMzbz3TtRYUiK1x4U] Player spawned
[2023-04-07T10:35:10Z INFO  ambient::client::wasm] [asset_loading_client] stderr: asset url can be accessed from client: ambient-assets:/assets/Cube.glb
[2023-04-07T10:35:10Z INFO  ambient_network::server] Allocating proxy endpoint
[2023-04-07T10:35:10Z INFO  ambient_network::server] Proxy sees this server as 88.89.173.176:9000
[2023-04-07T10:35:10Z INFO  ambient_network::server] Proxy allocated an endpoint, use `ambient join proxy-eu.ambient.run:9500` to join
```

**Comment by pierd** [comment]
> The `asset_loading` example yield following errors:
> 
> ```
> [2023-04-07T10:35:07Z INFO  ambient] Building Asset Loading
> [2023-04-07T10:35:10Z INFO  ambient] Done building Asset Loading
> [2023-04-07T10:35:10Z INFO  ambient::server] Creating server
> [2023-04-07T10:35:10Z INFO  ambient::server] Created server, running at 10.0.0.2:9000
> [2023-04-07T10:35:10Z INFO  ambient::server] Starting server
> [2023-04-07T10:35:10Z INFO  ambient_network::server] Connecting to proxy server
> thread 'tokio-runtime-worker' panicked at 'ambient_std::uncategorized::asset_url::AbsAssetUrl resource doesn't implement the load method', /Users/chaosprint/Dev/AmbientDev/Ambient/crates/asset_cache/src/lib.rs:443:9
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> [2023-04-07T10:35:10Z INFO  ambient_network::client] Connecting to server at 127.0.0.1:9000
> [2023-04-07T10:35:10Z INFO  ambient_network::server] [user_1MP7PFMzbz3TtRYUiK1x4U] Player spawned
> [2023-04-07T10:35:10Z INFO  ambient::client::wasm] [asset_loading_client] stderr: asset url can be accessed from client: ambient-assets:/assets/Cube.glb
> [2023-04-07T10:35:10Z INFO  ambient_network::server] Allocating proxy endpoint
> [2023-04-07T10:35:10Z INFO  ambient_network::server] Proxy sees this server as 88.89.173.176:9000
> [2023-04-07T10:35:10Z INFO  ambient_network::server] Proxy allocated an endpoint, use `ambient join proxy-eu.ambient.run:9500` to join
> ```
> 
> The `skinmesh` example also has similar errors and doesn't load the asset correctly:
> 
> ```
> [2023-04-07T10:36:57Z INFO  ambient] Building project
> [2023-04-07T10:36:58Z INFO  ambient] Done building project
> [2023-04-07T10:36:58Z INFO  ambient::server] Creating server
> [2023-04-07T10:36:58Z INFO  ambient::server] Created server, running at 10.0.0.2:9000
> [2023-04-07T10:36:58Z INFO  ambient::server] Starting server
> [2023-04-07T10:36:58Z INFO  ambient_network::server] Connecting to proxy server
> thread 'tokio-runtime-worker' panicked at 'ambient_std::uncategorized::asset_url::AbsAssetUrl resource doesn't implement the load method', /Users/chaosprint/Dev/AmbientDev/Ambient/crates/asset_cache/src/lib.rs:443:9
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> [2023-04-07T10:36:59Z INFO  ambient_network::client] Connecting to server at 127.0.0.1:9000
> [2023-04-07T10:36:59Z INFO  ambient_network::server] [user_EuTV1tnSC2nJqEu2MnYYF] Player spawned
> [2023-04-07T10:36:59Z INFO  ambient_network::server] Allocating proxy endpoint
> [2023-04-07T10:36:59Z INFO  ambient_network::server] Proxy sees this server as 88.89.173.176:9000
> [2023-04-07T10:36:59Z INFO  ambient_network::server] Proxy allocated an endpoint, use `ambient join proxy-eu.ambient.run:9554` to join
> ```
Thanks.

Yeah, I knew that I forgot about something. I’ll do more testing and fix it on Tuesday.

File: crates/network/src/client.rs [hunk 1]
```diff
@@ -231,7 +231,11 @@ impl ElementComponent for GameClientView {
                             let game_client =
                                 GameClient::new(conn, Arc::new(create_rpc_registry()), game_state.clone(), client_info.user_id);
 
-                            game_state.lock().world.add_resource(self::game_client(), Some(game_client.clone()));
+                            let world = &mut game_state.lock().world;
```
Comment (philpax) [inline crates/network/src/client.rs:234]: You may find `GameClient::with_physics_world` useful here (especially to disambiguate which world you're working with)

**Comment by philpax** [comment]
Seems reasonable to me! 🚀

File: app/src/server/wasm.rs [hunk 2]
```diff
@@ -55,11 +51,7 @@ pub fn initialize(world: &mut World, project_path: PathBuf, manifest: &ambient_p
 
             if target == "client" {
                 let relative_path = path.strip_prefix(&build_dir)?;
-
-                let assets = world.resource(asset_cache());
-                let base_url = ProxyBaseUrlKey.try_get(assets).unwrap_or_else(|| ServerBaseUrlKey.get(assets));
-                let bytecode_url = AssetUrl::parse(&relative_path.to_string_lossy())?.resolve(&base_url)?.to_string();
-
+                let bytecode_url = format!("{}:/{}", ASSETS_PROTOCOL_SCHEME, relative_path.to_string_lossy());
```
[the remark this answers is not in the document]
Comment (pierd) [reply app/src/server/wasm.rs:54]: Yes, root of current project

**Review by FredrikNoren — APPROVED** [review]

File: crates/network/src/client.rs (hunk 1 above)
Comment (pierd) [reply crates/network/src/client.rs:234]: Nice helper. I tried to use it but in this case the resulting code is actually more complex. I'd have to:
- destructure `ServerInfo` (`content_base_url` is used in one place and `project_name` in another)
- have additional block with clone for `game_client`

**Pull request merged**